### PR TITLE
Implement Median

### PR DESCRIPTION
**Pull request by NicMcPhee**
This implementation of the Median problem was mostly done by @GregoryPetterson with some help from me. I then did some clean-up at the end so it was closer to the model we'd worked out for the Smallest problem.

### CODE REVIEW

File: packages/push/examples/median/main.rs [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+    fn median(&self) -> Output {
+        let Self(mut input) = self;
+        input.sort_unstable();
+        Output(input[1])
+    }
+}
+
+impl Distribution<Input> for Uniform<i64> {
+    fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> Input {
+        Input([self.sample(rng), self.sample(rng), self.sample(rng)])
```
Comment (JustusFluegel) [inline packages/push/examples/median/main.rs:52]: This is probably nicer with
```rs
std::array::from_fn(|| self.sample(rng))
```
( didn't know that existed until a while ago, should probably do the same change for smallest )

File: packages/push/examples/median/main.rs [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+fn compute_error(final_state: &PushState, penalty_value: i128, expected: i64) -> i128 {
+    final_state
+        .stack::<i64>()
+        .top()
+        .map_or(penalty_value, |answer| {
+            i128::from(*answer)
+                .saturating_sub(i128::from(expected))
+                .abs()
+        })
+}
```
Comment (JustusFluegel) [inline packages/push/examples/median/main.rs:186]: These are swapped in `smallest` - do we care enough to keep a consistent order?

File: packages/push/examples/median/main.rs [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+
+fn instructions() -> impl Iterator<Item = PushInstruction> {
+    let int_instructions = IntInstruction::iter()
+        // Restore this line to remove `Min` from the instruction set.
+        // .filter(|&i| i != IntInstruction::Min)
```
Comment (JustusFluegel) [inline packages/push/examples/median/main.rs:191]: Does this make sense in median?

File: packages/push/examples/smallest/main.rs [hunk 4]
```diff
@@ -111,8 +109,6 @@ fn main() -> Result<()> {
         .into_collection_generator(population_size)
         .sample(&mut rng);
 
-    ensure!(population.is_empty().not());
```
Comment (JustusFluegel) [inline packages/push/examples/smallest/main.rs:114]: Why was this removed?

**Review by JustusFluegel — APPROVED** [review]
Lgtm, added a few minor comments.

File: packages/push/examples/median/main.rs (hunk 2 above)
Comment (NicMcPhee) [reply packages/push/examples/median/main.rs:186]: Probably 😆

File: packages/push/examples/median/main.rs (hunk 3 above)
Comment (NicMcPhee) [reply packages/push/examples/median/main.rs:191]: It probably does. If you have both `min` and `max` then it's easier. If you don't have both, then you need some fancier stuff.

File: packages/push/examples/smallest/main.rs (hunk 4 above)
Comment (NicMcPhee) [reply packages/push/examples/smallest/main.rs:114]: I'm happy to put it back if you think it's useful. I think it was a leftover from early on when I was writing the first `main` examples, and then I just wanted to make sure that things were working. It seemed like a "leftover" to me, so I removed it.

Comment (JustusFluegel) [reply packages/push/examples/smallest/main.rs:114]: I think this was probably a useful check, at least it seemed that way to me - I would just leave it in.

File: packages/push/Cargo.toml [hunk 5]
```diff
@@ -45,6 +45,7 @@ macros = ["dep:push_macros"]
 workspace = true
 
 [[package.metadata.example_runner.examples.smallest]]
+[[package.metadata.example_runner.examples.median]]
```
Comment (JustusFluegel) [inline packages/push/Cargo.toml:48]: doesn't this override the old config for smallest? and below as well?

**Review by JustusFluegel — APPROVED** [review]
LGTM - feel free to merge if you want to @NicMcPhee :)
